### PR TITLE
Server certificate validation

### PR DESCRIPTION
**Pull request by tbingol**
Verify server identity on the router via Subject Alternative Names part of the certificate if the newly introduced ambry.http2PeerCertificateSanRegex parameter is set. 

If the parameter is set, server tries to match at least one of the subject alternative names to the corresponding regular expression. In the case of failure, router closes the channel established between the server.

### CODE REVIEW

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+              continue;
+            }
+            int numSanName = 0;
+            for (Object subjectAltName : subjectAltNames) {
+              numSanName++;
+              List<?> entry = (List<?>) subjectAltName;
+              if (entry == null || entry.size() < 2) {
+                continue;
+              }
+              Integer altNameType = (Integer) entry.get(0);
```
Comment (cgtz) [inline ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java:96]: What do you think about also having a config for expected alt name type vs checking all types of alternative names for a match?

**Review by cgtz — CHANGES_REQUESTED** [review]

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2ChannelPoolHandler.java [hunk 2]
```diff
@@ -45,26 +45,32 @@
...
     this.connectionInboundExceptionHandler = new ConnectionInboundExceptionHandler();
+    this.http2ClientMetrics = http2ClientMetrics;
   }
 
   @Override
   public void channelCreated(Channel ch) throws Exception {
     ChannelPipeline pipeline = ch.pipeline();
     SslHandler sslHandler = new SslHandler(sslFactory.createSSLEngine(host, port, SSLFactory.Mode.CLIENT));
     pipeline.addLast(sslHandler);
+    pipeline.addLast(
```
[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2ChannelPoolHandler.java]: Done

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+              String altName = (String) entry.get(1);
+              if (altNameType == null || altName == null) {
+                continue;
+              }
+
+              logger.info("(TDBG) validateSslConnection => SAN[" + numSanName + "] => type: " + altNameType + " name: " + altName + " Regex: " + http2PeerCertificateSanRegex);
```
[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java]: Removed the log line.

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Active " + ctx.channel().remoteAddress());
+    validateSslConnection(ctx);
+    super.channelActive(ctx);
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Inactive " + ctx.channel().remoteAddress());
```
[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java]: Done

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                continue;
+              }
+              Integer altNameType = (Integer) entry.get(0);
+              String altName = (String) entry.get(1);
+              if (altNameType == null || altName == null) {
+                continue;
+              }
+
+              logger.info("(TDBG) validateSslConnection => SAN[" + numSanName + "] => type: " + altNameType + " name: " + altName + " Regex: " + http2PeerCertificateSanRegex);
+              if (Pattern.matches(http2PeerCertificateSanRegex, altName)) {
```
[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java]: Done

[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java]: Done

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2ChannelPoolHandler.java [hunk 6]
```diff
@@ -45,26 +45,32 @@
...
+        new Http2PeerCertificateValidator(http2ClientConfig.http2PeerCertificateSanRegex,
```
[the remark this answers is not in the document]
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2ChannelPoolHandler.java]: Done

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java (hunk 1 above)
Comment (tbingol) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java:100]: I feel like that would be an overkill and also error prone due to requiring users to enter a consistent type. Though not very likely, newly added types in future could also cause breakage. If you feel strong about it I can add it though.

Comment (cgtz) [reply ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java:100]: Nah, I don't feel too strongly about this. Your reasoning makes sense for keeping the configs simple.

**Review by cgtz — APPROVED** [review]
LGTM!

File: ambry-server/src/main/java/com/github/ambry/server/AmbryMain.java [hunk 7]
```diff
@@ -45,8 +45,8 @@ public static void main(String[] args) {
           Utils.getObj(clusterMapConfig.clusterMapClusterAgentsFactory, clusterMapConfig,
               options.hardwareLayoutFilePath, options.partitionLayoutFilePath);
       logger.info("Bootstrapping AmbryServer");
-      VcrClusterAgentsFactory vcrClusterAgentsFactory = Utils.getObj(cloudConfig.vcrClusterAgentsFactoryClass);
-      ambryServer = new AmbryServer(verifiableProperties, clusterAgentsFactory, vcrClusterAgentsFactory,
+//      VcrClusterAgentsFactory vcrClusterAgentsFactory = Utils.getObj(cloudConfig.vcrClusterAgentsFactoryClass);
+      ambryServer = new AmbryServer(verifiableProperties, clusterAgentsFactory, (VcrClusterAgentsFactory) null,
```
Comment (zzmao) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryMain.java:49]: null? Looks like some debug code .

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-server/src/main/java/com/github/ambry/server/AmbryMain.java (hunk 7 above)
Comment (tbingol) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryMain.java]: Removed it though my local server would not start without this change. How does your server start?

Comment (zzmao) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryMain.java]: I recently relies on integration test to have vcrClusterAgentsFactory created. But `null` also work for local test.

**Pull request merged**